### PR TITLE
Added a fancy render system to enable things like stairs and redstone…

### PR DESCRIPTION
**Pull request by soundlogic2236**
… dust to be rendered properly in the visualization and multiblock preview systems.

### CODE REVIEW

File: src/main/java/vazkii/botania/common/lexicon/LexiconData.java [hunk 1]
```diff
@@ -335,7 +336,14 @@ public static void init() {
 				new PageRuneRecipe("13", ModRuneRecipes.recipeLustRune), new PageRuneRecipe("14", ModRuneRecipes.recipeGluttonyRune), new PageRuneRecipe("15", ModRuneRecipes.recipeGreedRune), new PageRuneRecipe("16", ModRuneRecipes.recipeSlothRune), new PageRuneRecipe("17", ModRuneRecipes.recipeWrathRune), new PageRuneRecipe("18", ModRuneRecipes.recipeEnvyRune), new PageRuneRecipe("19", ModRuneRecipes.recipePrideRune));
 
 		terrasteel = new BLexiconEntry(LibLexicon.BASICS_TERRASTEEL, categoryBasics);
-		terrasteel.setLexiconPages(new PageText("0"), new PageCraftingRecipe("1", ModCraftingRecipes.recipeTerraPlate), new PageText("2"), new PageMultiblock("4", ModMultiblocks.terrasteelPlate), new PageTerrasteel("3")).setIcon(new ItemStack(ModItems.manaResource, 1, 4));
+		terrasteel.setLexiconPages(new PageText("0"), new PageCraftingRecipe("1", ModCraftingRecipes.recipeTerraPlate), new PageText("2"), new PageMultiblock("4", ModMultiblocks.terrasteelPlate), new PageTerrasteel("3"), new PageMultiblock("5", new Multiblock() { {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/common/lexicon/LexiconData.java]: Remove your debugs.

File: src/main/java/vazkii/botania/client/core/handler/MultiblockRenderHandler.java [hunk 2]
```diff
@@ -98,21 +105,75 @@ private boolean renderComponent(World world, Multiblock mb, MultiblockComponent
...
+				catch(Exception e) {
+					comp.doFancyRender = false;
+				}
+                tessellator.draw();
+				blockRender.renderAllFaces = false;
+				blockRender.blockAccess = oldBlockAccess;
+			}
+			else {
+				int color = block.getRenderColor(meta);
+		        float red = (float)(color >> 16 & 255) / 255.0F;
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/MultiblockRenderHandler.java]: Fix your indentation.

File: src/main/java/vazkii/botania/client/core/handler/MultiblockBlockAccess.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		this.anchorZ = anchorZ;
+		this.hasBlockAccess = access!=null;
+	}
+	
+	/**
+	 * Returns the multiblock component for the coordinates, adjusted based on the anchor
+	 */
+	protected MultiblockComponent getComponent(int x, int y, int z) {
+		MultiblockComponent comp = multiblock.getComponentForLocation(
+				x - anchorX,
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/MultiblockBlockAccess.java]: Don't put these on multiple lines.

File: src/main/java/vazkii/botania/client/core/handler/MultiblockBlockAccess.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	/**
+	 * Updates the block access to the new parameters
+	 */
+	public void update(IBlockAccess access, Multiblock mb, int anchorX, int anchorY, int anchorZ) {
+		this.originalBlockAccess = access;
+		this.multiblock = mb;
+		this.anchorX = anchorX;
+		this.anchorY = anchorY;
+		this.anchorZ = anchorZ;
+		this.hasBlockAccess = access!=null;
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/MultiblockBlockAccess.java]: Spaces.

File: src/main/java/vazkii/botania/api/lexicon/multiblock/Multiblock.java [hunk 5]
```diff
@@ -149,5 +158,33 @@ public int getYSize() {
 	public int getZSize() {
 		return Math.abs(minZ) + Math.abs(maxZ) + 1;
 	}
+	
+	/**
+	 * Rebuilds the location cache
+	 */
+	public void updateLocationCache() {
+		locationCache.clear();
+		for(MultiblockComponent comp : components) {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/api/lexicon/multiblock/Multiblock.java]: No brackets.

File: src/main/java/vazkii/botania/api/lexicon/multiblock/Multiblock.java [hunk 6]
```diff
@@ -29,15 +31,21 @@
...
 	public int minX, minY, minZ, maxX, maxY, maxZ, offX, offY, offZ;
+	
+	public HashMap<List<Integer>, MultiblockComponent> locationCache = new HashMap<List<Integer>, MultiblockComponent>();
 
 	/**
 	 * Adds a multiblock component to this multiblock. The component's x y z
 	 * coords should be pivoted to the center of the structure.
 	 */
 	public void addComponent(MultiblockComponent component) {
+		if(getComponentForLocation(component.relPos.posX, component.relPos.posY, component.relPos.posZ)!=null) {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/api/lexicon/multiblock/Multiblock.java]: Spaces, no brackets.

**Comment by phantamanta44** [comment]
Er, I would use rebase instead of creating a new commit.

**Pull request closed without merging**